plugin/hpsa/hpsa.py

```diff
@@ -909,3 +933,223 @@ def volume_ident_led_clear(self, volume, flags=Client.FLAG_RSVD):
                 "Volume not found")
 
         return None
+
+    @_handle_errors
+    def batteries(self, search_key=None, search_value=None,
+                  flags=Client.FLAG_RSVD):
+        lsm_batteries = []
+        ctrl_all_show = self._sacli_exec(
+            ["ctrl", "all", "show", "config", "detail"])
+
+        for ctrl_name in ctrl_all_show.keys():
+            ctrl_data = ctrl_all_show[ctrl_name]
+            if 'Battery/Capacitor Count' not in ctrl_data or \
+               ctrl_data['Battery/Capacitor Count'] == 0:
+                continue
+
+            sys_id = ctrl_data["Serial Number"]
+
+            battery_type = Battery.TYPE_UNKNOWN
+            battery_status = _hp_battery_status_to_lsm(ctrl_data)
+
+            if ctrl_data["Cache Backup Power Source"] == "Capacitors":
+                battery_type = Battery.TYPE_CAPACITOR
+            elif ctrl_data["Cache Backup Power Source"] == "Batteries":
+                battery_type = Battery.TYPE_CHEMICAL
+            else:
+                raise LsmError(
+                    ErrorNumber.PLUGIN_BUG,
+                    "Got unknown battery type: %s for system %s" %
+                    (ctrl_data["Cache Backup Power Source:"], sys_id))
+
+            for counter in range(0, int(ctrl_data["Battery/Capacitor Count"])):
+
+                lsm_batteries.append(
+                    Battery(
+                        "%s_BAT_%d" % (sys_id, counter),
+                        "Battery %d of %s" % (counter, ctrl_name),
+                        battery_type, battery_status, sys_id, None))
+
+        return lsm_batteries
+
+    @_handle_errors
+    def volume_cache_info(self, volume, flags=Client.FLAG_RSVD):
+        """
+        Depend on command:
+            hpssacli ctrl slot=0 show config detail
+        """
+        flag_battery_ok = False
+        flag_ram_ok = False
+
+        (ctrl_num, array_num, ld_num) = self._cal_of_lsm_vol(volume)
+        ctrl_data = self._sacli_exec(
+            ["ctrl", "slot=%s" % ctrl_num, "show", "config", "detail"]
+            ).values()[0]
+
+        lsm_bats = self.batteries()
+        for lsm_bat in lsm_bats:
+            if lsm_bat.status == Battery.STATUS_OK:
+                flag_battery_ok = True
```

Comment:
Multiple capacitors can definitely be supported within a chassis (it's required for multiple controllers that need a power safe cache prior to Gen9), but you'd only have one per controller/system.

In Gen9, we went to a shared battery model (hedged our bets that we'd need it for nvdimms in Gen9, and we were right). As far as I've heard so far today from asking a few people, there aren't any chassis that support multiple batteries. But if there were in the future, the logic is such that you need a certain amount of "loads" in the battery for whatever hardware that is present in the server. If you don't have enough loads to cover all the hardware in your box that would need the power, you're not power safe. So, while we don't deal with situations where one battery is good and another is bad today, you can envision a world were we would. My assumption is that hpsa would enable write cache as long as there was enough power spread across the batteries in a system to cover a power outage. But, then I'd also expect the battery status to be reported as OK, so what you're doing here would be fine in this case.